src/main/java/net/spy/memcached/tls/TLSConnectionManager.java

```diff
@@ -0,0 +1,211 @@
+package net.spy.memcached.tls;
+
+import net.spy.memcached.compat.log.Logger;
+import net.spy.memcached.compat.log.LoggerFactory;
+
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.SSLEngine;
+import javax.net.ssl.SSLEngineResult;
+import javax.net.ssl.SSLEngineResult.HandshakeStatus;
+import javax.net.ssl.SSLException;
+import javax.net.ssl.SSLSession;
+import java.io.Closeable;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.SocketChannel;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+
+public class TLSConnectionManager implements Closeable {
+
+    private static final Logger LOG = LoggerFactory.getLogger(TLSConnectionManager.class);
+
+    private final SSLEngine sslEngine;
+
+    private SSLSession currentSession;
+
+    private ByteBuffer appOutBuffer; // Holds the data we are preparing to send out
+    private ByteBuffer appInBuffer; // Holds the data we have received and unwrapped
+    private ByteBuffer networkOutBuffer; // Holds the data we are sending across the wire
+    private ByteBuffer networkInBuffer; // Holds the data we have received from the wire
+
+    ExecutorService executor = Executors.newSingleThreadExecutor();
+
+    public TLSConnectionManager(SSLContext sslContext) {
+        this.sslEngine = sslContext.createSSLEngine();
+
+        configureSslEngine();
+    }
+
+    private void configureSslEngine() {
+        // We are the client, not the server
+        sslEngine.setUseClientMode(true);
+    }
+
+    private void initBuffers(SSLSession session) {
+        // allocateDirect() to keep all bytes contiguous in memory
+        appOutBuffer = ByteBuffer.allocateDirect(session.getApplicationBufferSize());
+        appInBuffer = ByteBuffer.allocateDirect(session.getApplicationBufferSize());
+        networkOutBuffer = ByteBuffer.allocateDirect(session.getPacketBufferSize());
+        networkInBuffer = ByteBuffer.allocateDirect(session.getPacketBufferSize());
+    }
+
+    public boolean doHandshake(SocketChannel socketChannel) throws IOException {
+        LOG.debug("%s - Beginning handshake.", socketChannel.getRemoteAddress());
+        try {
+            sslEngine.beginHandshake();
+            currentSession = sslEngine.getSession();
+            initBuffers(currentSession);
+
+            HandshakeStatus handshakeStatus = sslEngine.getHandshakeStatus();
+            while (!Thread.currentThread().isInterrupted()
+                    && handshakeStatus != HandshakeStatus.FINISHED
+                    && handshakeStatus != HandshakeStatus.NOT_HANDSHAKING) {
+                LOG.debug("%s - Handshake status: %s", socketChannel.getRemoteAddress(), handshakeStatus.name());
```

Comment:
We need the check for if debug logging is enabled here before calling LOG.debug here, since if debug logging isn't enabled it would actually still format the string for the log anyways before it checks that the logger has debug enabled, and string formatting is a slow operation